feat: tree index api prototype (#22491)
microsoft/FluidFramework: .changeset/stinky-shoes-melt.md

---
"fluid-framework": minor
---
---
"section": "tree"
---

> Export alpha index APIs

Reviewer: ```suggestion

New alpha APIs for indexing
```

> This exposes two new index APIs, `createSimpleTreeIndex` and `createIdentifierIndex`.

Reviewer: ```suggestion
SharedTree now supports indexing via two new APIs, `createSimpleTreeIndex` and `createIdentifierIndex`.
```

> `createSimpleTreeIndex` can be used to create a `SimpleTreeIndex` that indexes nodes based on their schema.

Reviewer: ```suggestion
`createSimpleTreeIndex` is used to create a `SimpleTreeIndex` which indexes nodes based on their schema.
```

> Depending on the schema, the user specifies which field to key the node on.

Reviewer: ```suggestion
The developer can specify the schema field on which to key the index.
```

Author: this suggestion is a little inaccurate because the user must specify the field for a particular schema or nodes of that schema won't be included in the index

maybe something more like:
```suggestion
To index nodes of a particular schema, the developer must specify the schema field which to key the nodes.
```

> `createIdentifierIndex` is used to create an `IdentifierIndex` which provides an efficient way for users to retrieve nodes using the node identifier.

Reviewer: ```suggestion
`createIdentifierIndex` is used to create an `IdentifierIndex` which provides an efficient way to retrieve nodes using the node identifier.
```

---
"fluid-framework": minor
---
---
"section": "tree"
---

New alpha APIs for indexing

> SharedTree now supports indexing via two new APIs, `createSimpleTreeIndex` and `createIdentifierIndex`.

Reviewer 2: Could we add an example of how to use it? 😁
